common/debug-interface/src/trace.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use crate::json_log::JsonLogEntry;
+use anyhow::Result;
+use chrono::{DateTime, Duration, Utc};
+use reqwest::Client;
+use serde_json::{json, Value};
+
+const TRACE_EVENT: &str = "trace_event";
+const TRACE_EDGE: &str = "trace_edge";
+const PAGING_SIZE: usize = 10000;
+
+pub struct LibraTraceClient {
+    client: Client,
+    addr: String,
+}
+
+impl LibraTraceClient {
+    /// Create LibraTraceClient from a valid socket address.
+    pub fn new<A: AsRef<str>>(address: A, port: u16) -> Self {
+        let client = Client::new();
+        let addr = format!("http://{}:{}", address.as_ref(), port);
+
+        Self { client, addr }
+    }
+
+    pub async fn get_libra_trace(
+        &self,
+        start_time: DateTime<Utc>,
+        duration: Duration,
+    ) -> Result<Vec<(String, JsonLogEntry)>> {
+        let start = start_time.format("%FT%T%.3fZ").to_string();
+        let end = (start_time + duration).format("%FT%T%.3fZ").to_string();
+        let response = self
+            .client
+            .get(&format!("{}/_search?scroll=1m", self.addr))
+            .json(&json!(
+            {
+                "size": PAGING_SIZE,
+                "query": {
+                    "bool": {
+                        "must": [
+                            { "term": { "name":   "libra_trace"}}
+                        ],
+                        "filter": [
+                            { "range": { "@timestamp": { "gte": start, "lte": end }}}
+                        ]
+                    }
+                }
+            }
+            ))
+            .send()
+            .await?;
+
+        let response: Value = response.json().await?;
+
+        let mut id = response["_scroll_id"].clone();
```

Comment:
I think this will panic if response does not have _scroll_id.
You might want to use some method that returns Option (like .get) and return error if _scroll_id is not found, instead of panic

Reply:
Square brackets can also be used to index into a value in a more concise way. This returns Value::Null in cases where get would have returned None. response["_scroll_id"] will return Value::Null if not found I think. But we can make it return error.

Reply:
Ok, if you confirmed that `[]` does not panic, but return value::Null then we are good